core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -772,6 +772,12 @@ private[spark] class TaskSetManager(
   private[scheduler] def markPartitionCompleted(partitionId: Int): Unit = {
     partitionToIndex.get(partitionId).foreach { index =>
       if (!successful(index)) {
+        if (speculationEnabled) {
```

Comment:
IIUC in this case no task in this taskSet actually successfully finishes, it's another task attempt from another taskSet for the same stage that succeeded. In stead of changing this code path, I'd suggest we have another flag to show whether any task succeeded in current taskSet, and if no task have succeeded, skip L987.

WDYT @squito ?

Reply:
yeah that is sort of what I was suggesting -- but I was thinking rather than just a flag, maybe we separate out `tasksSuccessful` into `tasksCompletedSuccessfully` (from this taskset) and `tasksNoLongerNecessary` (from any taskset), perhaps with better names.  If you just had a flag, you would avoid the exception from the empty heap, but you still might decide to enable speculation prematurely as you really haven't finished enough for `SPECULATION_QUANTILE`: https://github.com/apache/spark/blob/a381bce7285ec30f58f28f523dfcfe0c13221bbf/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L987